### DISCUSSION TITLE
Fronius Modbus TCP with SML driver

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#19022] [Q&A] [answered]
@gemu2015 Can you give any hints as to how to structure the meter definition for Fronius Modbus TCP? I'm using the following manual for figuring it out:
<img width="599" alt="Screen Shot 2023-07-03 at 13 51 40" src="https://github.com/arendst/Tasmota/assets/79770016/de9ee87d-e744-4071-8809-79aff268eb2a">

Seems like TCP requests are significantly different than RTU.

A standard RTU request over TCP
```
>M 1
+1,[192.168.3.38],m,0,502,modbus,1,10,01039C8F
1,010304SSss@i0:10,PV1 Voltage,V,PV1_Voltage,1
```
Gives
```
13:53:45.809 > 01 03 9c 8f 00 02 db b0 
13:53:45.944 : 01 00 00 00 00 03 01 83 0b 
13:53:46.790 > 01 03 9c 8f 00 02 db b0 
13:53:46.970 : 01 00 00 00 00 03 01 83 0b 
13:53:47.871 > 01 03 9c 8f 00 02 db b0 
13:53:48.257 : 01 00 00 00 00 03 01 83 0b 
13:53:48.903 > 01 03 9c 8f 00 02 db b0 
13:53:48.959 : 01 00 00 00 00 03 01 83 0b 
```
And a TCP message:
`+1,[192.168.3.38],m,0,502,modbus,1,10,00010000000301039C8F` with `00 01` (transaction ID) `00 00` (Protocol ID) `00 03` (Length) `01` (Unit ID )`03` (Function code) `9C 8F` (Register address) gives:
```
14:17:37.124 > 00 01 00 00 00 03 01 03 9c 8f 00 02 88 55 
14:17:38.083 > 00 01 00 00 00 03 01 03 9c 8f 00 02 88 55 
14:17:39.092 > 00 01 00 00 00 03 01 03 9c 8f 00 02 88 55 
14:17:40.110 > 00 01 00 00 00 03 01 03 9c 8f 00 02 88 55 
```

Note that manual does not have CRC in TCP request.

**Comment by gemu2015** [comment]
the tcp modbus protocol is fixed,  it is automatically generated from the modbus transmit sequence as needed (yes without crc).
you can not use the above example 2
you should  simply use the same structure as for serial modbus.

**Comment by pkkrusty** [comment]
OK made some progress with it.  Getting serial responses. With meter definition `+1,[192.168.3.23],m,0,502,modbus,1,10,r01039C870013` I get:
```
11:07:27.170 > 01 03 9c 87 00 13 9a 7e 
11:07:27.546 : 01 00 00 00 00 29 01 03 26 04 cf 01 95 01 9d 01 9d ff fe a8 ac a7 94 a7 bc 64 fa 62 ca 5b 68 ff 
11:07:27.588 : fe 77 6a ff ff 13 89 ff fe 77 7d ff ff d5 d0 
```
Data starts after MBAP header of `01 00 00 00 00 29 01 03 26`

How do I structure the metrics?

I've tried `1,010326UUuu@i0:100,AC Current,A,AC_Current,2` and `1,010000000029010326UUuu@i0:100,AC Current,A,AC_Current,2` and neither works.

**Reply by gemu2015** [reply]
the metrics should be equivalent to serial TCP mode.
keep in mind that the dump mode is the data returned from TCP meter.
this data is converted to serial format before decoding. (header strapped)

so try:

1,0103UUuu@i0:100,AC Current,A,AC_Current,2

**Comment by pkkrusty** [comment, marked as answer]
`1,0103UUuu@i0:100,AC Current,A,AC_Current,2` did not work.  still returning 0. Normal RTU structure works correctly with `1,010304UUuu@i0:100,AC Current,A,AC_Current,2` on a meter definition of `+1,3,m,0,502,modbus,1,10,01039C87`. Data returned for Modbus TCP is `01 00 00 00 00 07 01 03 04 03 ec 01 4a`

*edit: fixed in latest dev version

**Reply by Sfre12** [reply]
Hey pkkrusty.
I saw this topic and hope you can help me out. (I am a newby at SML decoding)
I am using a ESP32 and a "fronius syno 5" over TCP
Already found some SML script that is working for me with the most datas.

>M 1
+1,[192.168.67.119],m,0,502,Fronius,1,10,r01039C870015
1,=so3,128
1,01032aUUuu@i0:100,AC Current,A,AC_Current,2
1,01032ax28UUuu@i0:100,Frequency,Hz,Frequency,2
1,01032ax24UUuu@i0:100,Output Power,W,Output_Power,0
1,01032ax36UUuu@i0:100,Reaktiv Power,VAR,VAR,1
1,01032ax2UUuu@i0:100,Phase 1 Current,A,Phase_1_Current,2
1,01032ax4UUuu@i0:100,Phase 2 Current,A,Phase_2_Current,2
1,01032ax6UUuu@i0:100,Phase 3 Current,A,Phase_3_Current,2
1,01032ax16UUuu@i0:100,Phase 1 Voltage,V,Phase_1_Voltage,1
1,01032ax18UUuu@i0:100,Phase 2 Voltage,V,Phase_2_Voltage,1
1,01032ax20UUuu@i0:100,Phase 3 Voltage,V,Phase_3_Voltage,1
1,01032ax36UUuu@i0:1,Temperature,C,Temperature,0
#

But now I have some problems. The Temperature is not correct at "ax36"
and also I did not find how to get other registers. I tried to extend the amount of registers from Hex 0015 to Hex 0022
(+1,[192.168.67.119],m,0,502,Fronius,1,10,r01039C870022)
but after doing that, no more data`s will be displayed. all values are "0"
Can you please guide me how to get other registers from my fronius syno ?
I want to get the energy production and the temperatures.
e.g. here are my register adresses from fronius

40094 | 40095 | 2 | R | 0x03 | WH | AC Lifetime   Energy production | uint32 | Wh | WH_SF | may be   imprecise
40096 | 40096 | 1 | R | 0x03 | WH_SF | AC   Lifetime Energy production scale factor | sunssf |   |   |  
40097 | 40097 | 1 | R | 0x03 | DCA | DC   Current value | uint16 | A | DCA_SF | Not   supported if multiple DC inputs; current can be found 
40098 | 40098 | 1 | R | 0x03 | DCA_SF | Scale   factor | sunssf |   |   | Not   supported if multiple DC inputs.
40099 | 40099 | 1 | R | 0x03 | DCV | DC   Voltage value | uint16 | V | DCV_SF | Not   supported if multiple DC inputs; voltage can be found in Multiple MPPT model
40100 | 40100 | 1 | R | 0x03 | DCV_SF | Scale   factor | sunssf |   |   | Not   supported if multiple DC inputs.
40101 | 40101 | 1 | R | 0x03 | DCW | DC   Power value | int16 | W | DCW_SF | Total   power of all DC inputs
40102 | 40102 | 1 | R | 0x03 | DCW_SF | Scale   factor | sunssf |   |   |  
40103 | 40103 | 1 | R | 0x03 | TmpCab | Cabinet   Temperature | int16 | °   C | Tmp_SF | Not   supported
40104 | 40104 | 1 | R | 0x03 | TmpSnk | Coolant   or Heat Sink Temperature | int16 | °   C | Tmp_SF | Not   supported
40105 | 40105 | 1 | R | 0x03 | TmpTrns | Transformer   Temperature | int16 | °   C | Tmp_SF | Not   supported
40106 | 40106 | 1 | R | 0x03 | TmpOt | Other   Temperature | int16 | °   C | Tmp_SF | Not   supported
40107 | 40107 | 1 | R | 0x03 | Tmp_SF | Scale   factor | sunssf |   |   | Not   supported

Best regards
Stephan

**Reply by pkkrusty** [reply]
Yeah, I noticed that the temperature is often not correct. I believe it also needs a scale factor. 

You need to adjust your data lines if you increase the meter readout. So if you read 34 bytes (0x22), you need to change the initial part of each line to read the data correctly. So you would start each line with 
```
1,010344UUuu@i0:100,AC Current,A,AC_Current,2
```
instead of 
```
1,01032aUUuu@i0:100,AC Current,A,AC_Current,2
```
See my latest production code for Fronius at https://github.com/arendst/Tasmota/discussions/19042#discussioncomment-8958016[https://github.com/arendst/Tasmota/discussions/19042#discussioncomment-8958016](url)

You can add individual registers to be read by adding them to the meter definition...for example, if you want `40094 | 40095 | 2 | R | 0x03 | WH | AC Lifetime Energy production | uint32 | Wh | WH_SF | may be imprecise` you would add `01039C9E` to 
```
M 1
+1,[192.168.67.119],m,0,502,Fronius,1,10,r01039C870015
1,=so3,128
```
so it would look like
```M 1
+1,[192.168.67.119],m,0,502,Fronius,1,10,r01039C870015,01039C9E
1,=so3,128
```
and then add a line to reference that like
```
1,010304UUuuUUuu@i1:1,Lifetime Prod,Wh,Life_Prod,3
```

**Reply by Sfre12** [reply]
Thanks, will try that!

**Reply by Sfre12** [reply]
Hey PKkrusty.
I spend this evening to check if I can get the datas.
I still want to know how it is working. From the docs on github, about decoder, I did not understand.

I tried a lot of things but got no success. So during testing I figure out I need more input ;-)

`1,010304UUuuUUuu@i1:1,Lifetime Prod,Wh,Life_Prod,3` can you explain the `04` Why using this 04?

To check data on Modbus I use QModMaster 

![image](https://github.com/arendst/Tasmota/assets/79339733/9966f4a0-8e64-49a2-bf72-4bff735fda08)

Do you know a better free windows tool to read the modbus register for testing?

**Reply by Sfre12** [reply]
OK, some news. I checked a lot of things, and figure out, the fronius documentation is not correct (or I am not able to read)
All tests I did is with a "Fronius SYMO 5.0-3-M"

Right know I got the "life time production" register...
my script:
`>D
>B
=>sensor53 r
>BS
=>sensor53 r
>M 1
+1,[192.168.67.119],m,0,502,Fronius,1,10,01039CFC
1,=so3,128
1,010304UUuuUUuu@i0:1000,Lifetime Production,kWh,Life_Prod,2
#`
and the output.
>D
>B
=>sensor53 r
>BS
=>sensor53 r
>M 1
+1,[192.168.67.119],m,0,502,Fronius,1,10,01039CFC
1,=so3,128
1,010304UUuuUUuu@i0:1000,Lifetime Production,kWh,Life_Prod,2
#'
And here is the output
![image](https://github.com/arendst/Tasmota/assets/79339733/e90e0e6a-d470-4eaa-84b8-4a3348d52c21)

Today the converter is working and I was able to do some more tests.
I tried to read the registers as single registers.
This ist working. Finally the docs from Fronius are not correct!

'>D
>B
=>sensor53 r
>BS
=>sensor53 r
>M 1
; ,   IPADRESS     , ,,Port,Label7n,?,??,Register,Register,Register,Register,Register,Register,Register,Register,Register,Register   
;1,[192.168.67.119],m,0,502,Fronius,1,10,40071   ,40085   ,40083   ,40071   ,40072   ,40073   ,40079   ,40080   ,40081   ,40076   ,40077   ,40078   ,40272   ,40273   ,40274   ,40188   ,40108
+1,[192.168.67.119],m,0,502,Fronius,1,10,01039C87,01039C95,01039C93,01039C88,01039C89,01039C8A,01039C8F,01039C90,01039C91,01039C8C,01039C8D,01039C8E,01039D50,01039D51,01039D52,01039CFC,01039CAC
1,=so3,256
1,010304UUuu@i0:100,AC Current,A,AC_Current,2
1,010304UUuu@i1:100,Frequency,Hz,Frequency,2
1,010304UUuu@i2:100,Output Power,W,Output_Power,0
1,010304UUuu@i3:100,Phase 1 Current,A,Phase_1_Current,2
1,010304UUuu@i4:100,Phase 2 Current,A,Phase_2_Current,2
1,010304UUuu@i5:100,Phase 3 Current,A,Phase_3_Current,2
1,010304UUuu@i6:100,Phase 1 Voltage,V,Phase_1_Voltage,1
1,010304UUuu@i7:100,Phase 2 Voltage,V,Phase_2_Voltage,1
1,010304UUuu@i8:100,Phase 3 Voltage,V,Phase_3_Voltage,1
1,010304UUuu@i9:100,Phase 1 Voltage,V,Phase_1_Voltage,1
1,010304UUuu@i10:100,Phase 2 Voltage,V,Phase_2_Voltage,1
1,010304UUuu@i11:100,Phase 3 Voltage,V,Phase_3_Voltage,1
1,010304UUuu@i12:100,DC Current,A,DC_Current,2
1,010304UUuu@i13:100,DC Voltage,V,DC_Voltage,1
1,010304UUuu@i14:100,DC Power,W,DC_Power,0
1,010304UUuuUUuu@i15:1000,Gesamt PV Produktion,kWh,ges_pv_prod,3
1,010304UUuu@i16:1,Status Code,,status,0
#'
![image](https://github.com/arendst/Tasmota/assets/79339733/22f13803-6cf3-4aac-9cfb-1a9c3ae261d2)

**Comment by pkkrusty** [comment]
Latest issue:

Fronius inverters use dynamic scale factors for certain registers, so ideally I would be able to perform math before sending info. I see in documentation that 
```
Perform arithmetic (+,-,*,/) on the measured data. Use # before a number to designate a constant value 
e.g. 1,=m 3+4+5/#3 @100,Voltage L1+L2+L3/3,V,Volt_avg,2 to sum results of decoder entries 3,4,5 and divide by 3 (average)
```
The scale factor is a signed integer resulting in -2, -1 or 0, which is then applied to the measured value (i.e. Voltage register reads 22345, scale factor -2, so 22345*10**-2=223.45)

Math operations look like they are limited to the four basics, without pow(), or bitwise operators. Is this true?

Second issue: I read 16 registers in a row, then create all my descriptor lines from that one return. Thus all my descriptor lines look like:
```
1,01032aUUuu@i0:100,AC Current,A,AC_Current,2
1,01032ax2UUuu@i0:100,Phase 1 Current,A,Phase_1_Current,2
1,01032ax4UUuu@i0:100,Phase 2 Current,A,Phase_2_Current,2
1,01032ax6UUuu@i0:100,Phase 3 Current,A,Phase_3_Current,2
```
So how would I define which descriptor lines I want to use? They are all number 0, with an offset. Do I need to re-write my meter description to break up the register reads into separate commands?

Ideally, I would be able to do something like:

`1,=m 2*(#10**5) @1,Curr Phs 1,A,curr_phs_1,2`

Or, as the scaling factor can use these negative numbers, could the descriptor line scaling factor optionally refer to another descriptor line? i.e. instead of `1,01032ax4UUuu@i0:-2,Phase 2 Current,A,Phase_2_Current,2` I could put `1,01032ax4UUuu@i0:4,Phase 2 Current,A,Phase_2_Current,2` and it would pull the value from descriptor line #4? But that would limit you to descriptor lines 2-9.  Maybe a 10 offset, so entering 11-99 lets you choose descriptor line 1-89.

**Reply by gemu2015** [reply]
descriptor math is meant for simple purposes only like adding 3 phases.

use script math instead.

each descriptor line can be accessed with sml[x] (x=1 ... n), so  simply do like that

```
>S
res=sml[1]*sml[3]*pow(10)
```

and then show result in >W and send the result in >J

**Comment by pkkrusty** [comment]
Issue 1: Noticing that my definition only works when returning <=21 bytes.
```
+1,[192.168.3.38],m,0,502,bouch,1,10,r01039C870015
```
works correctly, but once I try to return more (`+1,[192.168.3.38],m,0,502,bouch,1,10,r01039C870016`) the dump shows correct data returned, but values do not populate the GUI or JSON.

Issue 2: Returning 21 bytes in meter definition and descriptor referring to bytes 38 and 39 returns bad data, even though dump shows good data. Should be -2.

https://github.com/arendst/Tasmota/assets/79770016/81959862-3673-45df-91eb-8241108c2e7b


```
10:54:24.840 : 4a 01 00 00 00 2d 01 03 2a 04 75 01 77 01 78 01 86 ff fe 9d bc 9f 60 9f 4c 5b 54 5a 1e 5d c0 ff 
10:54:24.882 : fe 69 6e ff ff 13 87 ff fe 69 6e ff ff (00 00)<-byte 36/37 (ff fe)<-byte 38/39 27 10 
10:54:25.638 > 01 03 9c 87 00 15 1a 7c 
10:54:26.050 : 99 6a 00 00 00 2d 01 03 2a 04 84 01 7d 01 7c 01 8b ff fe 9d f8 9f 38 9f 10 5b 68 5a 1e 5d d4 ff 
10:54:26.091 : fe 6a ea ff ff 13 87 ff fe 6a ea ff ff 00 00 ff fe 27 10 
10:54:26.637 > 01 03 9c 87 00 15 1a 7c 
10:54:26.849 : af ae 00 00 00 2d 01 03 2a 04 84 01 7d 01 7c 01 8b ff fe 9d f8 9f 38 9f 10 5b 68 5a 1e 5d d4 ff 
10:54:26.891 : fe 6a ea ff ff 13 87 ff fe 6a ea ff ff 00 00 ff fe 27 10 
10:54:27.602 > 01 03 9c 87 00 15 1a 7c 
10:54:27.994 : d6 07 00 00 00 2d 01 03 2a 04 87 01 7e 01 7d 01 8c ff fe 9e 48 9f 4c 9f 4c 5b 72 5a 32 5d de ff 
10:54:28.035 : fe 6b 1c ff ff 13 87 ff fe 6b 1c ff ff 00 00 ff fe 27 10 
10:54:28.628 > 01 03 9c 87 00 15 1a 7c 
10:54:28.843 : 03 23 00 00 00 2d 01 03 2a 04 87 01 7e 01 7d 01 8c ff fe 9e 48 9f 4c 9f 4c 5b 72 5a 32 5d de ff 
10:54:28.885 : fe 6b 1c ff ff 13 87 ff fe 6b 1c ff ff 00 00 ff fe 27 10 
10:54:29.600 > 01 03 9c 87 00 15 1a 7c 
10:54:29.818 : d4 92 00 00 00 2d 01 03 2a 04 87 01 7e 01 7d 01 8c ff fe 9e 48 9f 4c 9f 4c 5b 72 5a 32 5d de ff 
10:54:29.859 : fe 6b 1c ff ff 13 87 ff fe 6b 1c ff ff 00 00 ff fe 27 10 
```

**Reply by gemu2015** [reply]
dump and decoder use different default buffer sizes
try to enlarge serial buffer for decoder, (see special options)
1,so3,128

**Reply by pkkrusty** [reply]
Documentation states that special options use format M,=so3,128 but when I do that, the line is ignored.

This is for use per descriptor line? i.e. 
```
+2,[192.168.3.23],m,0,502,mrcyshp,1,10,r01039C870015
2,=so3,128,01032aUUuu@i0:1,AC Current,A,AC_Current,0
```
Default buffer size for each section is what?

**Reply by gemu2015** [reply]
no special options are additional lines in descriptor
simply add this line
1,=so3,128
default input buffer is 48
default dump buffer is 128

**Reply by pkkrusty** [reply]
Ah ok, works like this:
```
+1,[192.168.3.38],m,0,502,bouch,1,10,r01039C870015
1,=so3,128
1,01032aUUuu@i0:1,AC Current,A,AC_Current,0
...
```

Thanks

**Comment by pkkrusty** [comment]
@gemu2015 on the descriptor line, `<txPeriod>	Period to repeat the transmission of commands to the meter (optional). Number of 100ms increments (n * 100ms).` often is set to 10, thus 1 second. Does this mean that every line address is requested each second? Are there buffer or speed issues to worry about with this? For example, if I have 40 lines of 2x 16-bit registers to request, that is 1280 bits, plus header and CRC, is 3840 bits. At 9600bps, that should take 400ms plus any overhead or delays between transmissions.  But also there may be buffer limits with the home screen gui and json. Do you know these off-hand?

**Reply by gemu2015** [reply]
at 10 each entry is requested every second.
in case you request 40 words per request this should fit with 1 second at 9600 baud
you also have to insure that your serial buffer is large enough to fit this request (header + 80 bytes + crc)

no limits with home screen gui and json

**Reply by pkkrusty** [reply]
So all registers are concatenated into one big list? So if I have a descriptor line like ...01039F3E,01039F10,01039F2A... the send is all together with one header/CRC?

**Reply by gemu2015** [reply]
no, each entry is sent separately and after each request the response is evaluated.
in the above case you only request the default 2 words per request
you may also request more than 2 words with the raw request like you did in the example above.

**Comment by 2000ede** [comment]
@pkkrusty :

Hello pkkrusty,
Do you have other report variables such as the meter reading?
or DC values?

Thanks!

>M 1
+1,[192.169.27.219],m,0,502,Fronius,1,10,r01039C870015
1,=so3,128
1,01032aUUuu@i0:100,AC Current,A,AC_Current,2
1,01032ax28UUuu@i0:100,Frequency,Hz,Frequency,2
1,01032ax24UUuu@i0:100,Output Power,W,Output_Power,0
1,01032ax36UUuu@i0:100,Reaktiv Power,VAR,VAR,1
1,01032ax2UUuu@i0:100,Phase 1 Current,A,Phase_1_Current,2
1,01032ax4UUuu@i0:100,Phase 2 Current,A,Phase_2_Current,2
1,01032ax6UUuu@i0:100,Phase 3 Current,A,Phase_3_Current,2
1,01032ax16UUuu@i0:100,Phase 1 Voltage,V,Phase_1_Voltage,1
1,01032ax18UUuu@i0:100,Phase 2 Voltage,V,Phase_2_Voltage,1
1,01032ax20UUuu@i0:100,Phase 3 Voltage,V,Phase_3_Voltage,1
;1,01032ax8SSss@i0:1,Curr Scale Fctr,SF,Curr_SF,0
;1,01032ax22SSss@i0:1,Vltg Scale Fctr,SF,Vltg_SF,0
;1,01032ax26SSss@i0:1,Pwr Scale Fctr,SF,Pwr_SF,0
;1,01032ax30SSss@i0:1,Freq Scale Fctr,SF,Freq_SF,0
;1,01032ax38SSss@i0:1,Temp Scale Fctr,SF,Temp_SF,0
#

**Reply by pkkrusty** [reply]
I seem to remember having trouble finding the correct registers. I would need to set up my test rig to cycle through all the registers and see what comes up. Probably not going to do that any time soon. Do you have any Fronius documents with all the listed registers?

**Reply by Sfre12** [reply]
Hey 200ede.
here are two docs I found at fronius page.
best regards
Stephan
[Gen24_Primo_Symo_Inverter_Register_Map_Int&SF.xlsx](https://github.com/arendst/Tasmota/files/14836732/Gen24_Primo_Symo_Inverter_Register_Map_Int.SF.xlsx)
[Smart_Meter_Register_Map_Int&SF.xlsx](https://github.com/arendst/Tasmota/files/14836733/Smart_Meter_Register_Map_Int.SF.xlsx)

**Comment by euro910** [comment]
Hi
did you have success on connecting the symo with a working script?
best regards Stefan

**Reply by pkkrusty** [reply]
I have been pulling data from Symo units for years now with no problems.

**Reply by euro910** [reply]
Good Morning,
i ment pushing to symo, emulating the smartmeter ip 
as i have no physical connection from smartmeter to the symo, sadly

**Reply by pkkrusty** [reply]
Are you wanting to control output dynamically? I haven't ever written to registers on a Symo unit, mostly because i can't find an accurate source for register numbers. If you could find accurate registers, I think it's possible.

**Reply by euro910** [reply]
yes, that´s what i expected to be able to
but it´s seems to be not as easy as i thought
also someone remarks that only the gen24 are controllable through modbus tcp
not the old symo inverter

**Reply by Sfre12** [reply]
Hi all.
I’ve now implemented this with Home Assistant and can control my Symo 5 to regulate the output power.
At the time, I created a document for it because I had spent a long time searching for a suitable solution.
If you’re interested, I can dig up the document again.
Regards
Stephan

**Reply by pkkrusty** [reply]
That would be interesting for me to see which registers you found.

**Reply by Sfre12** [reply]
[Fronius_Symo_Home_Assistant_Modbus_Doku.pdf](https://github.com/user-attachments/files/26481184/Fronius_Symo_Home_Assistant_Modbus_Doku.pdf)
Hi, unfortently it is in german, but you can use Translation Tools. 
Hopefully you will find what you need. 
Regards 
Stephan

**Reply by euro910** [reply]
perfect thank you Stephan
i will give it a try with the ottelo tasmota SML image
and Script from https://github.com/CherAlban/tasmota-sml-modbus-gen24-domoticz-ha-script?tab=readme-ov-file  
this is gen24 proved, i will compare it with your document (de is fine, greetings from Mosel)

**Reply by pkkrusty** [reply]
I tried yesterday based on that document and the excel file listed above and the registers match with offset -1. However register writes via Modbus TCP does not seem to work. I opened an issue with @gemu2015 who developed the driver to see if he can give some insight.
